torch/utils/data/dataloader.py

```diff
@@ -1191,30 +1191,36 @@ def _next_data(self):
                         self._workers_status[data.worker_id] = False
                     else:
                         self._mark_worker_as_unavailable(data.worker_id)
+                    # Given worker is exhausted, use _try_put_index to put data by looking for other available workers
                     self._try_put_index()
                     continue
 
+            worker_id = self._task_info[idx][0]
             if idx != self._rcvd_idx:
                 # store out-of-order samples
+                self._try_put_index(worker_id)
                 self._task_info[idx] += (data,)
             else:
                 del self._task_info[idx]
+                self._try_put_index(worker_id)
                 return self._process_data(data)
 
-    def _try_put_index(self):
+    def _try_put_index(self, worker_queue_idx=None):
         assert self._tasks_outstanding < self._prefetch_factor * self._num_workers
 
         try:
             index = self._next_index()
         except StopIteration:
             return
-        for _ in range(self._num_workers):  # find the next active worker, if any
-            worker_queue_idx = next(self._worker_queue_idx_cycle)
-            if self._workers_status[worker_queue_idx]:
-                break
-        else:
-            # not found (i.e., didn't break)
-            return
+
+        if (worker_queue_idx is None) or (not self._workers_status[worker_queue_idx]):
+            for _ in range(self._num_workers):  # find the next active worker, if any
+                worker_queue_idx = next(self._worker_queue_idx_cycle)
+                if self._workers_status[worker_queue_idx]:
+                    break
+            else:
+                # not found (i.e., didn't break)
+                return
 
```

Comment:
for modularity, you can create another small function called pick_worker and then place this logic there.